arbiter-engine/src/examples/token_minter.rs

```diff
@@ -0,0 +1,315 @@
+}
+
+impl TokenAdmin {
+    // TODO: I don't think we should pass in a client like this, probably, doing it
+    // for testing purposes. Also using RevmMiddleware for testing purposes,
+    // although this strategy should never be deployed
+    /// Creates a new token admin with the given identifier.
+    pub fn new(client: Arc<RevmMiddleware>) -> Self {
+        Self {
+            id: TOKEN_ADMIN_ID.to_owned(),
+            token_data: HashMap::new(),
+            tokens: None,
+            client,
+        }
+    }
+
+    /// Adds a token to the token admin.
+    pub fn add_token(&mut self, token_data: TokenData) {
+        self.token_data.insert(token_data.name.clone(), token_data);
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct TokenData {
+    pub name: String,
+    pub symbol: String,
+    pub decimals: u8,
+    pub address: Option<Address>,
+}
+
+/// Used as an action to ask what tokens are available.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub enum TokenAdminQuery {
+    /// Get the address of the token.
+    AddressOf(String),
+
+    /// Mint tokens.
+    MintRequest(MintRequest),
+}
+
+/// Used as an action to mint tokens.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct MintRequest {
+    /// The token to mint.
+    pub token: String,
+
+    /// The address to mint to.
+    pub mint_to: Address,
+
+    /// The amount to mint.
+    pub mint_amount: u64,
+}
+
+#[async_trait::async_trait]
+impl Strategy<Message, MessageOrTx> for TokenAdmin {
+    #[tracing::instrument(skip(self), fields(id = %self.id))]
+    async fn sync_state(&mut self) -> Result<()> {
+        debug!("Syncing state for `TokenAdmin`.");
+        for token_data in self.token_data.values_mut() {
+            let token = ArbiterToken::deploy(
+                self.client.clone(),
+                (
+                    token_data.name.clone(),
+                    token_data.symbol.clone(),
+                    token_data.decimals,
+                ),
+            )
+            .unwrap()
+            .send()
+            .await
+            .unwrap();
+            token_data.address = Some(token.address());
+            self.tokens
+                .get_or_insert_with(HashMap::new)
+                .insert(token_data.name.clone(), token.clone());
+            trace!("Deployed token: {:?}", token);
+        }
+        Ok(())
+    }
+
+    #[tracing::instrument(skip(self, event), fields(id = %self.id))]
+    async fn process_event(&mut self, event: Message) -> Vec<MessageOrTx> {
+        trace!("Processing event for `TokenAdmin` {:?}.", event);
+        if self.tokens.is_none() {
+            error!("There were no tokens to deploy! You must add tokens to the token admin before running the simulation.");
+        }
+        if event.to == self.id {
+            let query: TokenAdminQuery = serde_json::from_str(&event.data).unwrap();
+            trace!("Got query: {:?}", query);
+            match query {
```

Comment:
also thing that it's a good pattern to match on message type. Do you think in the future we could abstract some of the logic into helpers with descriptive to make it intuitive? or do you think thats not really important?

Reply:
Other way to do it is to break it into multiple behaviors, each being simple and dealing with one type of event. I think this is also a valid pattern, however I did it this way because I was having to debug multi-agent stuff. I plan to revisit this example and display two ways to do it to see which UX seems better. However, the end user should always have their own choice be valid.